dbms/src/Interpreters/Join.cpp

```diff
@@ -617,79 +624,108 @@ void NO_INLINE insertFromBlockImplTypeCaseWithLock(
     ConstNullMapPtr null_map,
     Join::RowRefList * rows_not_inserted_to_map,
     size_t stream_index,
-    Arena & pool)
+    Arena & pool,
+    bool enable_fine_grained_shuffle)
 {
     KeyGetter key_getter(key_columns, key_sizes, collators);
-    std::vector<std::string> sort_key_containers;
-    sort_key_containers.resize(key_columns.size());
-    size_t segment_size = map.getSegmentSize();
-    /// when inserting with lock, first calculate and save the segment index for each row, then
-    /// insert the rows segment by segment to avoid too much conflict. This will introduce some overheads:
-    /// 1. key_getter.getKey will be called twice, here we do not cache key because it can not be cached
-    /// with relatively low cost(if key is stringRef, just cache a stringRef is meaningless, we need to cache the whole `sort_key_containers`)
-    /// 2. hash value is calculated twice, maybe we can refine the code to cache the hash value
-    /// 3. extra memory to store the segment index info
-    std::vector<std::vector<size_t>> segment_index_info;
-    if (has_null_map && rows_not_inserted_to_map)
+    std::vector<std::string> sort_key_containers(key_columns.size());
+
+    if (enable_fine_grained_shuffle)
```

Comment:
Can we use a new function to build hash map with fine_grained_shuffle?
such as
```
if (insert_concurrency > 1)
{
    if (enable_fine_grained_shuffle)
       insertFromBlockImplTypeCaseWithFineGrainedShuffle<STRICTNESS, KeyGetter, Map, false>(map, rows, key_columns, key_sizes, collators, stored_block, null_map, rows_not_inserted_to_map, stream_index, pool); 
    else
        insertFromBlockImplTypeCaseWithLock<STRICTNESS, KeyGetter, Map, false>(map, rows, key_columns, key_sizes, collators, stored_block, null_map, rows_not_inserted_to_map, stream_index, pool);
}
else
{
    insertFromBlockImplTypeCase<STRICTNESS, KeyGetter, Map, false>(map, rows, key_columns, key_sizes, collators, stored_block, null_map, rows_not_inserted_to_map, stream_index, pool);
}
```

Reply:
Nice suggestion, will change it .

Reply:
Done.